<h1>PIC 16 Final Exam, Part II</h1>
This part of the exam is optional*. If you are not satisfied with your ability to complete Part I, you may complete <i>one or more</i> of the three options below. If you do so, your score for the options you complete will be factored into your total score for the exam. The weights are: 20% for Option 1, 20% for Option 2, and 30% for Option 3.

For instance, let’s say you were to score 70 points out of 100 on Part I. <br>
You also attempted Part II, Option 1, scoring 35 out of the 40 points available for that option. <br>
Finally, you somehow had time to complete Option 3, scoring 18 out of the 18 points available for that option. <br>
Then your grade for the exam would be:  70/100 x 50% + 35/40 x 20% + 20/20 x 30% = 82.5%. 

Regarding time: I wrote these questions over a week ago and attempted them again recently, so I remembered some of the concepts but definitely didn't have the solution code memorized. On this attempt, I even messed up a bit on Option 1 Problem 5 and crashed Jupyter during Option 2 Problem 1, so it wasn't perfectly smooth sailing. Still, each option took me about 20 minutes. I usually estimate that it will take most students about 3x as long as I do in these circumstances. So if you are fluent in the material from the first half of the course but didn't learn anything in the second part of the course, you could probably complete all three of these options during the exam and get a score of up to 70% for the exam. However, if you struggle through Part I and realize with only an hour left that it's not going well, you might only have time for one.

\* (unless you chose Part I Track B Option 3, in which case you must complete at least one of the options below)

<h2>Option 1</h2>
This option consists of 10 separate, rather short questions on material from the first few weeks of the course. If you appear to have made an attempt to complete it, your score (out of 40 points) will account for 20% of your grade for the exam. (The weight of Part I will be reduced accordingly.) If you do not intend for it to be included in your exam grade, please remove all Option 1 cells from this notebook before submitting.

<i>If code (such as a function) is provided in a separate cell, you should execute that cell before executing your solution cell.</i>

<b>1a. (4 pts)</b><br>
Write a function <tt>f</tt> that accepts arguments <tt>x</tt> and <tt>y</tt> and returns <br> $f(x,y) = \sin(x) e^{-xy} + 2\cos(10y)$. <br>
For full credit, avoid the use of <tt>pow</tt> and <tt>**</tt>.

In [1]:
from math import *
def f(x,y):
    return sin(x)*exp(-x*y) + 2*cos(10*y)

f(1,1)

-1.3685831824997927

<b>1b. (1 pt)</b><br>
Test <tt>f</tt> by evaluating it at $x = \pi$, $y = 1$ and printing the result, which should be <tt>-1.6781430581529049</tt>.

In [2]:
f(pi,1)

-1.6781430581529049

-1.6781430581529049

<b>2a. (3 pts)</b><br>
Write a function <tt>g</tt> that accepts a tuple <tt>z</tt> consisting of elements <tt>x</tt> and <tt>y</tt> (in that order) and returns both $f(x,y)$ and $2f(x,y)$ (in that order).

In [10]:
def g(z):
    print z[0], z[1]
    return f(z[0],z[1]), 2*f(z[0], z[1])
g((1,2))

1 2


(0.9300448376911521, 1.8600896753823042)

<b>2b. (2 pts)</b><br>
Test <tt>g</tt> by evaluating it with a tuple consisting of the values <tt>1</tt> and <tt>2</tt> (in that order) and assigning the results to variables <tt>a</tt> and <tt>b</tt>. Print <tt>a</tt> (0.930044837691) and <tt>b</tt> (1.86008967538).

In [11]:
a,b = g((1.0,2.0))
print a,b

1.0 2.0
0.930044837691 1.86008967538


<b>3. (4 pts)</b><br>
Write a <i>lambda</i> function <tt>h</tt> that computes the sum of its arguments (assuming they are numeric), regardless of how many there are. 

In [22]:
h = lambda *k: reduce(lambda x,y: x+y, list(k))

Test it by running the cell below. The result should be 15.

In [24]:
# DO NOT DELETE
h(1,2,3,4,5,6)

21

<b>4. (2 pts)</b><br>
Write a function or lambda function <tt>numUnique</tt> that returns the number of <i>unique</i> elements in its argument, which you can assume is a list. For maximum credit, the function should contain only one line of code.

In [25]:
numUnique = lambda x: len(set(x))

Test it by running the cell below. The result should be <tt>4</tt>.

In [26]:
# DO NOT DELETE
numUnique([1,2,1,2,3,2,4])

4

<b>5. (3 pts)</b><br>
Write a function or <i>lambda</i> function <tt>lists2dict</tt> that accepts two arguments, lists <tt>a</tt> and <tt>b</tt>, and returns a dictionary <tt>c</tt> composed of keys from <tt>a</tt> and corresponding values from <tt>b</tt>. For full credit, the function should contain only one line of code and avoid the use of <tt>range</tt>.

In [28]:
lists2dict = lambda a,b: {i:j for i,j in zip(a,b) }

Test it by running the cell below. The result should be <tt>{'a': 1, 'b': 2, 'c': 3}</tt>.

In [29]:
# DO NOT DELETE
a = ["a","b","c"]
b = [1,2,3]
lists2dict(a,b)

{'a': 1, 'b': 2, 'c': 3}

<b>6. (3 pts)</b><br>
Write a <i>lambda</i> function <tt>trimReduceReverse</tt> that accepts a list argument and performs the following:
<ul>
<li>eliminates the first and last elements,</li>
<li>removes any entries that are <tt>None</tt>, and</li>
<li>returns the remaining list in reverse order.</li>
</ul> <br>
For full credit, avoid the use of the <tt>reverse</tt> method, and the <tt>len</tt>, <tt>range</tt>, and <tt>reversed</tt> functions.

In [52]:
trimReduceReverse = lambda x: [i for i in x[-2:0:-1] if i!=None]

Test it by running the cell below. The result should be <tt>[5, 4, 3, 2]</tt>.

In [53]:
# DO NOT DELETE
a = [1,2,None,3,4,None,5,6]
print trimReduceReverse(a)

[5, 4, 3, 2]


<b>7. (2 pts)</b><br>
Without a <tt>for</tt> loop or any sort of Python comprehension, apply the function:

In [56]:
# Remember to run this cell
def f(x):
    return x**2

to all elements in a list

In [57]:
# This one, too
l = range(10)

and print the result, which should be <tt>[0, 1, 4, 9, 16, 25, 36, 49, 64, 81]</tt>.

In [61]:
import numpy as np
k = np.array(l)
f(k)

array([ 0,  1,  4,  9, 16, 25, 36, 49, 64, 81])

<b>8. (2 pts)</b><br>
Why do the following two examples seem to behave differently? A complete explanation involves the words "stack", "heap", "mutable", and "immutable", integer, list, object. Answer by editing the text beneath the examples.

In [62]:
x = 1
y = x
y += 1
print x
print y

1
2


In [17]:
x = [1]
y = x
y += [1]
print x
print y

[1, 1]
[1, 1]


<b>Response to 8</b> (Double click to edit)

In the first example, x is a variable on the (heap) that (refers to) an integer object on the (stack) and y is a variable on the (heap) that refers to (a different) integer object. Integer objects are (immutable), so when we try to increment y, y is incremented and x stays the same since y copies x (fill in the rest yourself)

In the second example, x is a variable on the (heap) that (refers to) a list object on the (stack) and y is a variable on the (heap) that refers to (the same) list object. Integer objects are (immutable), so when we try to append to y, the element is appended to the list on the stack which x also refers to (fill in the rest yourself)

<b>9. (4 pts)</b><br>
Consider the function:

In [63]:
# Remember to run this cell before running your solution cell
import random
def f():
    x = random.random()
    if x < 0.5:
        return x
    else:
        raise ValueError("Value was greater than 0.5")

Write code that executes <tt>f</tt>. If <tt>f</tt> returns a value, print it. Otherwise, if it raises a <tt>ValueError</tt>, do nothing (nothing should be printed, either).

In [88]:
try:
    print f()
except:
    pass

0.374109374332


<b>10. (10 pts)</b><br>
Write a class <tt>MutableInt</tt> such that the code:
<pre>
x = MutableInt(1)
y = x
print x, y
y += 1
print x, y
</pre>
produces:
<pre>
1 1
2 2
</pre>
In other words, instances of <tt>MutableInt</tt> are supposed to behave like integers except that they are <i>mutable</i> objects on the heap.

In [42]:
class MutableInt:
    MutableInt.num = 0
    def __init__(self, x):
        print "init"
        MutableInt.num = x

    
    def __add__(self, x):
        print "add"
        MutableInt.num  = MutableInt.num  + x
    
#     def __getitem__(self):
#         print "getitem"
#         return self.x
    
#     def __setitem__(self, k):
#         print "setitem"
#         self.x += k
#         return self.x
    
    def __str__(self):
        #print "str"
        return str(MutableInt.num)

Test your class by running the code below.

In [43]:
# DO NOT DELETE
x = MutableInt(1)
y = x
print x, y
y += 1
print x, y

init
1 1
add
2 None


<h2>Option 2</h2>
This option consists of 2 medium-length questions on material from the first few weeks of the course. If you appear to have made an attempt to complete it, your score (out of 45 points) will account for 20% of your grade for the exam. (The weight of Part I will be reduced accordingly.) If you do not intend for it to be included in your exam grade, please remove all Option 2 cells from this notebook before submitting.

<b>11. Pokemon Three Ways:</b><br>
Consider the class <tt>Pokemon</tt>. 

In [99]:
# Remember to run this cell
import math
class Pokemon(object):
    def __init__(self,species, level, favNo):
        self.species = species
        self.level = level
        self.favNo = favNo
        self.i = 0

<b>11a. (10 points)</b><br>
<b>I recommend saving your work frequently while working on this problem! "My code crashed Jupyter" is not a good excuse for lost work.</b> 

Write a subclass, <tt>Itermon</tt>, that can be iterated over using a <tt>for</tt> loop like:
<pre>
p = Itermon("Pikachu", 10, math.pi)
for field in p:
    print field
</pre>
to print the values of all three fields:
<pre>
Pikachu
10
3.14159265359
</pre>
For full credit, your subclass should <i>not</i> contain <tt>species</tt>, <tt>level</tt>, or <tt>favNo</tt> attributes/fields of its own; it should rely on those in the superclass.

In [104]:
class Itermon(Pokemon):
    def __init__(self, *x):
        super(Itermon, self).__init__(self, *x)
        
    def __iter__(self):
        self.iter =  tuple(self.species, self.level, self.favNo)
        return self.iter
    
    def next(self):
        if self.i == len(self.iter):
            raise StopIteration
        else:
            self.i+=1
            return self.iter[self.i-1]
        
            

Test your class using the code below. It should print the values in all three fields twice.

In [105]:
# DO NOT DELETE
p = Itermon("Pikachu", 10, math.pi)
for field in p:
    print field
for field in p:
    print field

TypeError: __init__() takes exactly 4 arguments (5 given)

<b>11b. (10 points)</b><br>
Write two more classes <tt>Itermon2</tt> and <tt>Itermon3</tt> that accomplish the same objective as in 11b, but each class should use a different technique. Of your three classes (in 11a and 11b), one should use a generator, one should use a generator expression, and one should use neither.

Test your classes using the code below. It should print the values in all three fields four times.

In [39]:
# DO NOT DELETE
p = Itermon2("Pikachu", 10, math.pi)
for field in p:
    print field
for field in p:
    print field
    
p = Itermon3("Pikachu", 10, math.pi)
for field in p:
    print field
for field in p:
    print field

Pikachu
10
3.14159265359
Pikachu
10
3.14159265359
Pikachu
10
3.14159265359
Pikachu
10
3.14159265359


<b>12. (15 points)</b><br>
Write a program that prompts the user for the name of a text file. If the text file is present in the same directory as this notebook, print the contents of the file. Otherwise, inform the user that the file is not present and prompt again. The program should terminate if the user types 'q', 'Q', 'quit', or 'Quit' instead of a filename. <br> A file <tt>test.txt</tt> is provided for your convenience. <br>
Example interaction:
<pre>
Enter the name of a file to display: test.txt
hi there
these are contents 
Enter the name of a file to display: notpresent.txt
There was a problem opening that file. Maybe it doesn't exist!
Enter the name of a file to display: q
</pre>

Enter the name of a file to display: test.txt
hi there
these are contents 
Enter the name of a file to display: notpresent.txt
There was a problem opening that file. Maybe it doesn't exist!
Enter the name of a file to display: q


<h2>Option 3</h2>
This option consists a single, multi-part GUI problem. If you appear to have made an attempt to complete it, your score (out of 20 points) will account for 30% of your grade for the exam. (The weight of Part I will be reduced accordingly.) If you do not intend for it to be included in your exam grade, please remove all Option 3 cells from this notebook before submitting.

<b>13. (20 points)</b><br>
Write a program that accomplishes the following objectives:
<ul>
<li>(3 pts) Create a window 300px x 300 px with a white background and the title (in the blue bar) "Spinner"</li>
<li>(3 pts) Draw a black vertical line of length 200 centered in the window</li>
<li>(3 pts) The vertical line should remain centered as the window is resized.</li>
<li>(3 pts) The line should rotate clockwise, remaining centered in the window.</li>
<li>(3 pts) Text above the line should count the number of seconds since the program began.</li>
<li>(3 pts) The line should undergo exactly (according to system time) one quarter revolution per second.</li>
<li>(2 pts) The text above the line should read  "0" and the line should be still until a button at the top left "Click to Start" is clicked </li>
</ul> <br>
This problem requires basic trigonometry. 
If $(x,y)$ is the point at angle $\theta$ on a circle of radius $r$ centered at $x_c, y_c$, then <br>
$(x-x_c) = r \cos(\theta)$ <br>
$(y-y_c) = r \sin(\theta)$ <br>
I've included some code to get you started.

In [1]:
from PyQt4.QtGui import QApplication, QWidget

def main():
    app = QApplication([])
    q = QWidget()
    q.show()
    app.exec_()

if __name__ == "__main__":
    main()